File: particles/Tutorial/Javascript/7_RefinementTypes/source/DisplayGreeting.js
/**
 * @license
 * Copyright 2019 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

'use strict';

/* global defineParticle */


defineParticle(({SimpleParticle, html}) => {

  const template = html`Hello, <span>{{num1}}</span>!
  <div slotid="inputSlot1"></div>`;

  return class extends SimpleParticle {
    get template() {
      return template;
    }

    // We need the person handle within shouldRender, so it has to be passed in.
    shouldRender({person1}) {
      // Here we check that the person is defined.
      return person1;
    }

    // Just like with shouldRender, we need access to person, so declare it needs to be passed in.
    render({person1}) {
      // We want the num from person to be interpolated into the template.
      console.log(person1);
      return {
        num1: person1 && person1.num || 0
      };
    }
  };
});

File: particles/Tutorial/Kotlin/4_Handles/DisplayGreeting.kt
/*
 * Copyright 2019 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.tutorials

import arcs.addressable.toAddress
import arcs.Handle
import arcs.Particle
import arcs.Singleton
import kotlin.native.internal.ExportForCppRuntime
import kotlin.native.Retain

/**
 * Sample WASM Particle.
 */
class DisplayGreetingParticle : Particle() {
    private val person = Singleton { DisplayGreeting_Person() }

    override fun getTemplate(slotName: String) = "Hello, <span>{{name}}</span>!"

    init {
        registerHandle("person", person)
    }

    override fun onHandleUpdate(handle: Handle) {
        this.renderOutput()
    }

    override fun populateModel(slotName: String, model: Map<String, Any?>): Map<String, Any?> {
        return model + mapOf(
            "name" to (person.get()?.name ?: "Human")
        )
    }
}

@Retain
@ExportForCppRuntime()
fun _newDisplayGreeting() = DisplayGreetingParticle().toAddress()
